docs/history-framework.md

```diff
@@ -193,7 +193,113 @@ sequenceDiagram
+
+#### Session start id plus index
+Create a CLUE document session record (Firestore document) at the beginning of working with each CLUE document. Store the history entries under this documentSession document with an index for the session. 
+The documentSession document would have a server timestamp that can be used to order them.
+
+We load all of the documentSession documents ordered by the serverTimestamp and their children history entries. 
+
+If the id of the documentSession is generated client side then the history entries don't need to wait.
+
+##### Problems
+We don't really know about overlapping entries here. 2 sessions could be active at the same time. The entries could include a server timestamp so we'd have a sense of this. We might be able to update the sessionDocument with an ending timestamp when the user closes the document, but that probably won't be reliable, since a network failure can break it.
+
+If one session is started, closed and other started very quickly the sessionDocument timestamps might be out of order.
+
+##### Benefits
+No waiting at the beginning of a session for a round trip to Firestore.
+Order of entries is in firestore 
+
+#### Other Notes
+
+We could use local computer timestamps to avoid the problem out of order events, but these are not accurate and in some schools can be years off.
+
+There might be an approach that does the "Session start id plus index" without using extra documents by putting this start time or id in each of the history entry documents.
+
+We could also put some session info the parent document itself, like a `numberOfSessions` to the document. So then a session index could be added to each history entry document. Then the history entry events could be ordered by the combination of this session index and a history entry index. 
+
+An alternative to querying the entries for the last index would be to store the last history entry in the parent document, but this means that document have to be updated with each new entry. And there is a limit of 1 sec per update. If you want to look into that approach more see:
+https://firebase.google.com/docs/firestore/solutions/aggregation
+
+### Serialization TODO:
+- [ ] See if we can remove this use of cloud functions to do the network document writes, I think the only reason to use them is for permissions, I'd guess we can create rules that would allow them to be written without the functions. We probably are going to have either do this or change the cloud function so it can support students creating parent documents.
+- [ ] fix ordering issue, currently we are using a server timestamp which gets out of order when the first entry is delayed as the document parent is fetched. The right approach I think is to store the last index in the document and then on each session load this index and start from there. This last index will also be useful when dealing with publishing documents.
```

Comment:
I don't think a cloud function would help in this case. We still can't trust the order the entries arrive at the server. So we have to have a client side mechanism to record this ordering. 

I forgot to update this TODO, I kind of ruled out this approach of storing the last index in the parent document because it requires the parent doc has to be updated on every history entry which will likely exceed the 1 update/s sometimes.